### ISSUE TITLE
Escaping is broken for $ during substitution in Vim

### ISSUE THREAD

**Issue by tejasvi** [codemirror/CodeMirror#6580]
![image](https://user-images.githubusercontent.com/45873379/106825134-384d5e00-66aa-11eb-974f-0fe722759f74.png)
gives
![image](https://user-images.githubusercontent.com/45873379/106825147-3d121200-66aa-11eb-9ddd-168f13ee1676.png)

As the case with Vim, `:%s/$/dollar` doesn't work either. However search using `/\$` works.

**Comment by benhormann** [comment]
You have found a bug in [this workaround](https://github.com/codemirror/CodeMirror/blob/5.59.2/keymap/vim.js#L5069).

The underlying issue relates to vim's use of [searchcursor](https://codemirror.net/doc/manual.html#addon_searchcursor) and zero-width matches. In [searchcursor.js line 242](https://github.com/codemirror/CodeMirror/blob/5.59.2/addon/search/searchcursor.js#L242-L254) there is code (workaround for vim?) that breaks zero-width matches at end of line boundary.

We can fix the use case above in vim.js, but it'll still fail when $ isn't at the end of the pattern (e.g. `(a|$)` or `$|b`). It will also fail for other assertion types. Example `:%s/(?<=buffe)r/d/` finds matches and moves the cursor, but doesn’t replace anything or show an error.
Here’s roughly what happens:
- searchcursor finds a match, returns the matched substring and position.
- vim does a regex replace on the substring. Starts again at position + 1.

With lookahead/lookbehind assertions, the pattern can be larger than the match substring. In these cases the replace method finds no match in the substring. In my example the substring is "r" but the pattern requires "buffer" to do the substitution.

I'm not sure which is more effort: 1) making vim.js work with searchcursor.js, or 2) removing the dependency and workarounds.
Can we change searchcursor to return an empty match if at the line boundary (and somehow avoid matching the same boundary over and over)?

---

TL;DR Here's a simple patch that'll *improve\** the workaround:
```patch
5069c5069
<           if (regexPart && regexPart[regexPart.length - 1] === '$') {
---
>           if (/(^|[^\\])(\\\\)*\$$/.test(regexPart)) {
```
@marijnh Would you accept such a pattern?

\* Doesn't fix when $ is mid-pattern, or the other assertions. E.g. `:s/\b/@/g` adds \@ at start of words but not at the end.

**Comment by marijnh** [comment]
I don't understand enough of vim, or even our vim bindings, to really provide a well-founded judgement here. So, uh, if this fixes some cases that were previously broken without breaking new cases, I guess I'm okay with it. (If possible, add a test case if you create a PR.)

**Comment by benhormann** [comment]
And I don't know enough about searchcursor (or whatever else depends on it). I'd say effort would be better spent designing/building a new vim module for CM6, then backport that to CM5 once complete enough. Well, that's the dream...

It won't break anything that isn't already broken, so I'll go ahead with a test and PR.
